crates/ruff_linter/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs

```diff
@@ -0,0 +1,620 @@
+                // The class/parent method's docstring can directly precede the def.
+                && !tracked_vars.follows_docstring
+                // Do not trigger when the def follows an if/while/etc...
+                && prev_indent_level.is_some_and(|prev_indent_level| prev_indent_level >= indent_level)
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !tracked_vars.follows_decorator
+            {
+                // E301
+                let mut diagnostic = Diagnostic::new(
+                    BlankLineBetweenMethods(line.line.preceding_blank_lines),
+                    token.range,
+                );
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist.line_ending().as_str().to_string(),
+                    locator.line_start(tracked_vars.last_non_comment_line_end),
+                )));
+
+                context.push_diagnostic(diagnostic);
+            }
+
+            if line.line.preceding_blank_lines < BlankLinesConfig::TOP_LEVEL
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !tracked_vars.follows_decorator
+                // Allow groups of one-liners.
+                && !(tracked_vars.follows_def
+                    && line
+                    .tokens_trimmed()
+                    .last()
+                    .map_or(false, |token| !matches!(token.kind(), TokenKind::Colon))
+                )
+                // Only trigger on non-indented classes and functions (for example functions within an if are ignored)
+                && indent_level == 0
+                // Only apply to functions or classes.
+                && is_top_level_token_or_decorator(token.kind)
+            {
+                // E302
+                let mut diagnostic = Diagnostic::new(
+                    BlankLinesTopLevel(line.line.preceding_blank_lines),
+                    token.range,
+                );
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist.line_ending().as_str().to_string().repeat(
+                        (BlankLinesConfig::TOP_LEVEL - line.line.preceding_blank_lines) as usize,
+                    ),
+                    locator.line_start(tracked_vars.last_non_comment_line_end),
+                )));
+
+                context.push_diagnostic(diagnostic);
+            }
+
+            if line.line.blank_lines > BlankLinesConfig::TOP_LEVEL
+                || (indent_level > 0 && line.line.blank_lines > BlankLinesConfig::METHOD)
+            {
+                // E303
+                let mut diagnostic =
+                    Diagnostic::new(TooManyBlankLines(line.line.blank_lines), token.range);
+
+                let chars_to_remove = if indent_level > 0 {
+                    line.line.preceding_blank_characters - BlankLinesConfig::METHOD
```

Comment:
Would it be possible to compute the blank lines ad-hoc? Storing them for every line seems expensive, especially because they're only needed for fixes.

Reply:
There isn't a way to go back to previous logical lines, right ?
Then the only option (afaik) would be to work with text and to go back until finding the end of the previous non-blank line. My understanding is that it would be slow, however if that is run only for fixes, then maybe it's fine.